packages/oauth-js/src/client.ts

```diff
@@ -0,0 +1,361 @@
+    } else {
+      this.authState = "ERROR";
+    }
+    removeTokenCookies(this.options);
+    this.storage.removeItem(STATE_DATA_KEY);
+  }
+
+  async refresh() {
+    const initialState = this.authState;
+    const lock = new Lock();
+
+    try {
+      this.authState = "UPDATING";
+
+      if (await lock.acquireLock(REFRESH_LOCK_KEY)) {
+        const refreshToken = getRefreshToken(this.options);
+        const response = await this.post("token", {
+          client_id: this.config.clientId,
+          refresh_token: refreshToken,
+          grant_type: "refresh_token",
+          redirect_uri: this.config.callbackUri,
+          scope: this.config.scope || "",
+        });
+
+        this.processResponse(response);
+      } else {
+        console.warn("Could not aquire refresh lock.");
+      }
+    } catch (error: any) {
+      console.log(error);
+      if (initialState !== "INIT") {
+        // refresh expected to fail in initial state
+      } else {
+        this.authState = "ERROR";
+      }
+    } finally {
+      await lock.releaseLock(REFRESH_LOCK_KEY);
+    }
+  }
+
+  async introspect() {
+    const token = getSessionToken(this.options);
+    const response = await this.post("token/introspect", {
+      client_id: this.config.clientId,
+      token_type_hint: "access_token",
+      token,
+    });
+
+    return await response.json();
+  }
+
+  async userinfo() {
+    const response = await this.get("userinfo");
+
+    return await response.json();
+  }
+
+  async getToken() {
+    if (this.shouldTokenRefresh()) {
+      try {
+        await this.refresh();
+      } catch (err) {}
+    }
+
+    const accessToken = getSessionToken(this.options);
+
+    if (!accessToken) {
+      // throw error
+      this.authState = "ERROR";
+    }
+
+    return accessToken;
+  }
+
+  private checkTokenLife() {
+    const tokenExpire = getTokenExpire(this.options);
+
+    if (tokenExpire && isTokenExpiring(tokenExpire)) {
+      this.refresh();
+    }
+  }
+
+  private startTokenWatch() {
+    const intervalTime = REFRESH_CHECK_INTERVAL * 1000;
+
+    this.stopTokenWatch();
+
+    this.timer = window.setInterval(() => {
+      this.checkTokenLife();
+    }, intervalTime);
+  }
+
+  private stopTokenWatch() {
+    if (this.timer) {
+      clearInterval(this.timer);
+      this.timer = null;
+    }
+  }
+
+  private processResponse(data: any) {
+    // TODO: fix data type
+    console.log("RESPONSE", data);
+
+    const user: any = {}; // getUserFromResponse(response); // TODO: fix user data
+    const sessionData = getSessionData(this.options);
+    sessionData.user = user;
+    saveSessionData(sessionData, this.options);
+
+    const storageAPI = getStorageAPI(this.options.useCookie);
+    const returnURL = storageAPI.getItem(LAST_PATH_KEY) || "/";
+
+    const appState = {
+      userData: user,
+      returnPath: returnURL,
+      authState: storageAPI.getItem(STATE_DATA_KEY),
+    };
+
+    storageAPI.removeItem(LAST_PATH_KEY);
```

Comment:
are these storage apis promises? just wondering if this processResponse should also be a promise, to ensure things like the locks aren't released until the response is fully processed. If they are not promises they can ignore, just thought I saw from the lock library it had integrations to local storage as promises

Reply:
No, the storage apis aren't promises